IQKeyboardManager/Categories/IQUIView+Hierarchy.m

```diff
@@ -89,35 +89,54 @@ -(UIViewController*)viewController
     return nil;
 }
 
--(UIViewController *)topMostController
+- (UIViewController *)topMostController
 {
-    NSMutableArray *controllersHierarchy = [[NSMutableArray alloc] init];
-    
-    UIViewController *topController = self.window.rootViewController;
-    
-    if (topController)
+    UIViewController *result = nil;
+    UIWindow *window = [[UIApplication sharedApplication] keyWindow];
+    if (window.windowLevel != UIWindowLevelNormal)
     {
-        [controllersHierarchy addObject:topController];
+        NSArray *windows = [[UIApplication sharedApplication] windows];
+        for(UIWindow *tmpWin in windows)
+        {
+            if (tmpWin.windowLevel == UIWindowLevelNormal)
+            {
+                window = tmpWin;
+                break;
+            }
+        }
     }
     
-    while ([topController presentedViewController]) {
-        
-        topController = [topController presentedViewController];
-        [controllersHierarchy addObject:topController];
+    id  nextResponder = nil;
+    UIViewController *appRootVC = window.rootViewController;
+    if (appRootVC.presentedViewController) {
+        nextResponder = appRootVC.presentedViewController;
+    } else {
+        UIView *frontView = [[window subviews] objectAtIndex:0];
+        nextResponder = [frontView nextResponder];
     }
     
-    UIResponder *matchController = [self viewController];
-    
-    while (matchController != nil && [controllersHierarchy containsObject:matchController] == NO)
-    {
-        do
-        {
-            matchController = [matchController nextResponder];
-            
-        } while (matchController != nil && [matchController isKindOfClass:[UIViewController class]] == NO);
+    if ([nextResponder isKindOfClass:[UINavigationController class]]){
+        UINavigationController *nav = (UINavigationController *)nextResponder;
+        result = nav.visibleViewController;
+    } else if ([nextResponder isKindOfClass:[UITabBarController class]]){
+        UITabBarController *tabbar = (UITabBarController *)nextResponder;
+        UIViewController *selectedVC = tabbar.selectedViewController;
+        if ([selectedVC  isKindOfClass:[UINavigationController class]]) {
+            UINavigationController *nav = (UINavigationController *)selectedVC;
+            result = nav.visibleViewController;
+        } else {
+            result = selectedVC;
+            while (result.presentedViewController) {
+                result = result.presentedViewController;
+            }
+        }
+    } else {
+        result = nextResponder;
+        while (result.presentedViewController) {
+            result = result.presentedViewController;
+        }
     }
-    
-    return (UIViewController*)matchController;
+    return result;
```

Comment:
If user has their own customized hierarchy to show something on screen then this may return wrong controller.